example_allocators/freelist_allocator/freelist_allocator.h

```diff
@@ -0,0 +1,52 @@
+/**********************************
+ * freelist_allocator.h
+ * Jeremy.Singer@glasgow.ac.uk
+ *
+ * This is a simple segregated freelist allocator.
+ * It mmaps a large buffer of num_pages pages,
+ * then constructs a linked list of LARGE-sized cells.
+ *
+ * When an alloc request occurs, we have three size
+ * classes we can use - SMALL, MEDIUM, and LARGE.
+ * If there is a empty cell available in the appropriate
+ * freelist, we return this cell.
+ * If there are no cells available, we try to
+ * grab a cell from a larger freelist to replenish
+ * our freelist, and return one of these cells.
+ *
+ * When a dealloc request occurs, we know the size
+ * of the cell so we can prepend the cell onto the
+ * appropriate freelist.
+ *
+ * Allocated cells have their sizes encoded in the
+ * corresponding cell capability - this means we
```

Comment:
Other than passing the wrong capability back to `dealloc`, can users do anything else to interfere with this?

Reply:
hmmm, that's interesting - what are you imagining?

Reply:
In a sense I'm imagining that they can't do anything else bad other than passing back the wrong capability :) I wonder if we can/need to verify/validate the capability in any way? I think the answer is that we can't usefully verify/validate it beyond just using it and seeing if we violate the capability. I might be wrong!